### PR TITLE
Add support for NOT Equals != operator

### PR DESCRIPTION
**Pull request by alexander-schranz**
Add not equals operator.

And fix NOT IN operator.

--> own pull request fixed in #20 

Currently the negative operators do not work as expected, when try to filter out all which matches a specific value they are still shown aslong as they have another value which should not be the case.

Example in SEAL:

| Id | Tags        | Expected | Current     |  Why?
|----|-------------|----------|-------------|-----------
| 1  | Tech<br> UI |          |   X         | Has one none -> logic need to be inverted
| 2  | UI<br>UX    |          |   X         | Has one none -> logic need to be inverted
| 3  | Tech<br>UX  |    X    |   X          | 
| 4  |             |    X    |              | Empty currently filter out -> logic need to be inverted

The query is `tags != 'UI'` and the expected result is `3` and `4`, but the result is: `1`, `2`, `3`. Same currently appears for the `NOT IN` filter I adopted the existing test cases there as they still are returning which should filtered out.

Also adopted the test cases for the expected result.

## Solution

The current SQL looks like this:

```sql
// != example
SELECT COUNT() OVER() AS totalHits, d.document
FROM documents d WHERE d.id IN (
    SELECT document FROM multi_attributes_documents mad
    INNER JOIN multi_attributes ma
        ON ma.attribute=:dcValue1
        AND ma.id = mad.attribute
    WHERE ma.string_value != 'Backoffice'
)
ORDER BY d.firstname ASC LIMIT 20

// NOT IN example
SELECT COUNT() OVER() AS totalHits, d.document
FROM documents d WHERE d.id IN (
    SELECT document FROM multi_attributes_documents mad
    INNER JOIN multi_attributes ma
        ON ma.attribute=:dcValue1
        AND ma.id = mad.attribute
    WHERE ma.string_value NOT IN ('Backoffice', 'Project Managemen')
)
ORDER BY d.firstname ASC LIMIT 20
```

But to get the expected result we would require to change for `NOT` statements (NOT IN and !=) the current SQL which looks like this:

```sql
// != example
SELECT COUNT() OVER() AS totalHits, d.document
FROM documents d WHERE d.id NOT IN (  // <--- replace IN with NOT IN
    SELECT document FROM multi_attributes_documents mad
    INNER JOIN multi_attributes ma
        ON ma.attribute=:dcValue1
        AND ma.id = mad.attribute
    WHERE ma.string_value = 'Backoffice' // <--- replace != with =
)
ORDER BY d.firstname ASC LIMIT 20

// NOT IN example
SELECT COUNT() OVER() AS totalHits, d.document
FROM documents d WHERE d.id NOT IN (  // <--- replace IN with NOT IN
    SELECT document FROM multi_attributes_documents mad
    INNER JOIN multi_attributes ma
        ON ma.attribute=:dcValue1
        AND ma.id = mad.attribute
    WHERE ma.string_valu IN ('Backoffice', 'Project Managemen') // <--- replace != with =
)
ORDER BY d.firstname ASC LIMIT 20
```

## TODO

 - [ ] We should maybe add test fixtures somebody which have maybe no `departments` currently empty multi value is not tested.

### CODE REVIEW

File: tests/Functional/SearchTest.php [hunk 1]
```diff
@@ -156,18 +231,10 @@ public static function inFilterProvider(): \Generator
                     'id' => 3,
                     'firstname' => 'Alexander',
                 ],
-                [
-                    'id' => 4,
-                    'firstname' => 'Jonas',
```
Comment (alexander-schranz) [inline tests/Functional/SearchTest.php:161]: `Jonas` is part of `Project Management` and should not be shown.

File: tests/Functional/SearchTest.php [hunk 2]
```diff
@@ -156,18 +231,10 @@ public static function inFilterProvider(): \Generator
...
-                ],
                 [
                     'id' => 1,
                     'firstname' => 'Sandra',
                 ],
-                [
-                    'id' => 2,
-                    'firstname' => 'Uta',
```
Comment (alexander-schranz) [inline tests/Functional/SearchTest.php:169]: `Uta` is part of `Backoffice` and should not be shown

File: src/Internal/Search/Searcher.php [hunk 3]
```diff
@@ -404,9 +404,14 @@ private function handleFilterAstNode(Node $node, array &$whereStatement): void
...
             if (\in_array($node->attribute, $this->engine->getIndexInfo()->getMultiFilterableAttributes(), true)) {
-                $whereStatement[] = $documentAlias . '.id IN (';
+                // @see https://github.com/loupe-php/loupe/pull/14
+                $previousOperator = $node->operator;
+                $operator = $node->operator->getMultiValueOperator();
+                $node->operator = $operator;
+                $whereStatement[] = $documentAlias . '.id ' . ($node->operator !== $previousOperator ? 'NOT IN' : 'IN') . ' (';
                 $whereStatement[] = $this->createSubQueryForMultiAttribute($node);
                 $whereStatement[] = ')';
+                $node->operator = $previousOperator;
```
Comment (alexander-schranz) [inline src/Internal/Search/Searcher.php:414]: I'm not happy with this code but maybe you have a better idea.

File: tests/Functional/SearchTest.php [hunk 4]
```diff
@@ -15,6 +15,81 @@ class SearchTest extends TestCase
 {
     use FunctionalTestTrait;
 
+    public static function equalFilterProvider(): \Generator
+    {
+        yield '= on multiple attribute' => [
```
Comment (Toflar) [inline tests/Functional/SearchTest.php:20]: Do you know if applying a single value operator like `=` or `!=` works on array-value attributes in other search engines? Does this work in MeiliSearch for example?

**Review by Toflar — COMMENTED** [review]
I think we should split fixing the `NOT IN` filter and adding support for `!=` (and applying those on multi attributes (to be checked if that's normal first)) in 2 PRs because they are not really related.

File: tests/Functional/SearchTest.php (hunk 4 above)
Comment (alexander-schranz) [reply tests/Functional/SearchTest.php:20]: @Toflar yes `=` and `!=` on multi values did work in all other engines yet which are supported by SEAL.

Comment (Toflar) [reply tests/Functional/SearchTest.php:20]: Okay, then let's add support for this in a separate PR. We should first fix the `NOT IN` logic and then add support for single comparisons on multi attributes. We should then also check what does happen when I apply `>=` on multi attributes.

Comment (alexander-schranz) [reply tests/Functional/SearchTest.php:20]: I can tackle the split at the evening. I could check what the case for `>=` and `<=` is in other engines on multi values and give you an update.

Comment (alexander-schranz) [reply tests/Functional/SearchTest.php:20]: @Toflar Fix for `NOT IN`: https://github.com/loupe-php/loupe/pull/19

Comment (alexander-schranz) [reply tests/Functional/SearchTest.php:20]: @Toflar `>=` and `<=` works on multi values also added additional test cases and the did work on currently all supported engines. https://github.com/schranz-search/schranz-search/pull/226

**Comment by Toflar** [comment]
Thanks, Alex!
